### PR TITLE
Task based refactor

### PR DESCRIPTION
**Pull request by markharley**
## Why are these changes needed?
* Refactor to facilitate improvement of time-series fitting.
* Delegate task specific logic to a Task sub-class, simplifying the AutoML class.
* Introduce a TimeSeriesDataset class for encapsulation of time-series relevant data operations.
* Introduce a TimeSeriesEstimator parent class to standardise time-series models.

## Related issue number
N/A

## Checks

- [X] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [X] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by gmdiana-hershey** [comment]
Was looking at the failed build and noticed that there were some import issues. Python 3.6 doesn't include data classes, but there is a backport. Adding that to the python 3.6 build script would solve that issue. Also saw that importing holidays fails. That's not part of the [test] requirements in setup.py (see https://github.com/markharley/FLAML/blob/fafd6524c9466d8ab9c9049bb6102c34cd5819ec/setup.py#L91).

File: flaml/ml.py [hunk 1]
```diff
@@ -1,710 +1,628 @@
...
-    from sklearn.preprocessing import label_binarize
-
-    classes = np.unique(y_true)
-    y_true_binary = label_binarize(y_true, classes=classes)
-
-    curve_x, curve_y = {}, {}
-    for i in range(len(classes)):
-        curve_x[i], curve_y[i], _ = curve_func(y_true_binary[:, i], y_pred_proba[:, i])
-    return curve_x, curve_y
+# !
```
Comment (int-chaos) [inline flaml/ml.py:1]: Entire file is shown as edited, which makes it hard to review. Make sure EOL (LF vs CRLF) is kept consistent, unless we are actually changing the EOL.

File: flaml/default/suggest.py [hunk 2]
```diff
@@ -1,259 +1,260 @@
...
-        X = model._preprocess(X)
-        hyperparams = hyperparams and model.params
-
-        class AutoMLTransformer:
-            def transform(self, X):
-                return model._preprocess(dt.transform(X))
-
-        transformer = AutoMLTransformer()
-        return hyperparams, estimator_class, X, y, transformer, dt.label_transformer
+import numpy as np
```
Comment (int-chaos) [inline flaml/default/suggest.py:1]: Entire file is shown as edited, which makes it hard to review. Make sure EOL (LF vs CRLF) is kept consistent, unless we are actually changing the EOL.

File: flaml/model.py [hunk 3]
```diff
@@ -1,2369 +1,1767 @@
...
-        os.dup2(self.null_fds[1], 2)
-
-    def __exit__(self, *_):
-        # Re-assign the real stdout/stderr back to (1) and (2)
-        os.dup2(self.save_fds[0], 1)
-        os.dup2(self.save_fds[1], 2)
-        # Close the null files
-        os.close(self.null_fds[0])
-        os.close(self.null_fds[1])
+# !
```
Comment (int-chaos) [inline flaml/model.py:1]: Entire file is shown as edited, which makes it hard to review. Make sure EOL (LF vs CRLF) is kept consistent, unless we are actually changing the EOL.

File: website/docs/Use-Cases/Task-Oriented-AutoML.md [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# Task Oriented AutoML
+# GenericTask Oriented AutoML
```
Comment (sonichi) [inline website/docs/Use-Cases/Task-Oriented-AutoML.md:1]: Is this and other changes from Task->GenericTask in the documentation a mistake?

File: flaml/nlp/huggingface/training_args.py [hunk 5]
```diff
@@ -1,9 +1,7 @@
 import argparse
 from dataclasses import dataclass, field
 
-from ...data import (
-    NLG_TASKS,
-)
+from flaml.automl.task import NLG_TASKS
```
Comment (liususan091219) [inline flaml/nlp/huggingface/training_args.py:4]: The `...data` is meant to be kept as the relative path to be consistent with the other files. Is there any reasons to change it to the absolute path?

File: flaml/nlp/huggingface/data_collator.py [hunk 6]
```diff
@@ -6,7 +6,11 @@
 )
 from collections import OrderedDict
 
-from flaml.data import TOKENCLASSIFICATION, MULTICHOICECLASSIFICATION, SUMMARIZATION
+from flaml.automl.task import (
+    MULTICHOICECLASSIFICATION,
+    TOKENCLASSIFICATION,
+    SUMMARIZATION,
+)
```
Comment (liususan091219) [inline flaml/nlp/huggingface/data_collator.py:13]: Here it was meant to be updated to the relative path as well.

File: flaml/nlp/huggingface/utils.py [hunk 7]
```diff
@@ -2,12 +2,12 @@
 from itertools import chain
 import numpy as np
 
-from ...data import (
-    SUMMARIZATION,
-    SEQREGRESSION,
+from flaml.automl.task import (
```
Comment (liususan091219) [inline flaml/nlp/huggingface/utils.py:5]: The ...data is meant to be kept as the relative path to be consistent with the other files. Is there any reasons to change it to the absolute path?

File: flaml/nlp/huggingface/training_args.py (hunk 5 above)
Comment (liususan091219) [inline flaml/nlp/huggingface/training_args.py:4]:
```suggestion
from ...automl.task import NLG_TASKS
```

File: flaml/nlp/huggingface/data_collator.py [hunk 8]
```diff
@@ -6,7 +6,11 @@
 )
 from collections import OrderedDict
 
-from flaml.data import TOKENCLASSIFICATION, MULTICHOICECLASSIFICATION, SUMMARIZATION
+from flaml.automl.task import (
```
Comment (liususan091219) [inline flaml/nlp/huggingface/data_collator.py:9]:
```suggestion
from ...automl.task import (
```

File: flaml/nlp/huggingface/utils.py (hunk 7 above)
Comment (liususan091219) [inline flaml/nlp/huggingface/utils.py:5]:
```suggestion
from ...automl.task import (
```

File: flaml/nlp/utils.py [hunk 9]
```diff
@@ -1,13 +1,12 @@
 from typing import Dict, Any
 import numpy as np
 
-from ..data import (
-    SUMMARIZATION,
-    SEQREGRESSION,
+from flaml.automl.task import (
```
Comment (liususan091219) [inline flaml/nlp/utils.py:4]:
```suggestion
from ...automl.task import (
```

File: flaml/data.py [hunk 10]
```diff
@@ -7,47 +7,21 @@
...
-SEQCLASSIFICATION = "seq-classification"
-MULTICHOICECLASSIFICATION = "multichoice-classification"
-TOKENCLASSIFICATION = "token-classification"
-CLASSIFICATION = (
-    "binary",
-    "multiclass",
-    "classification",
-    SEQCLASSIFICATION,
-    MULTICHOICECLASSIFICATION,
+from flaml.automl.task import (
```
Comment (liususan091219) [inline flaml/data.py:10]:
```suggestion
from .automl.task import (
```

File: flaml/default/estimator.py [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import sklearn.ensemble as ensemble
 from functools import wraps
-from flaml.data import CLASSIFICATION
+from flaml.automl.task import CLASSIFICATION
```
Comment (liususan091219) [inline flaml/default/estimator.py:3]:
```suggestion
from .automl.task import CLASSIFICATION
```

**Comment by liususan091219** [comment]
> Was looking at the failed build and noticed that there were some import issues.
> 
> * Python 3.6 doesn't include data classes, but there is a backport that could be added to the build script (and setup).
> * Also saw that importing holidays fails. That's not part of the [test] requirements in setup.py (see https://github.com/markharley/FLAML/blob/fafd6524c9466d8ab9c9049bb6102c34cd5819ec/setup.py#L91).
> * It sounds like new tests were added for time series tasks? Should the requirements in [ts_forecast] and [forecast] also be included in [test] in that case?

The holidays are only used in time series forecasting which is optional. Make sure the import statement is invoked only under the time series environment. See https://github.com/microsoft/FLAML/blob/main/flaml/automl.py#L1096 as an example for nlp (another optional environment).

**Review by qingyun-wu — COMMENTED** [review]
Would it be more reasonable to put the "nlp" and "time_series" folders into the "automl" folder?

File: flaml/ml.py (hunk 1 above)
Comment (markharley) [reply flaml/ml.py]: Hey. Thanks for the review. Yeah, looks like one of our editors wasn't happy with CRLF. I've reverted the line endings back to DOS

File: flaml/default/suggest.py (hunk 2 above)
Comment (markharley) [reply flaml/default/suggest.py:1]: Hey. Thanks for the review. Yeah, looks like one of our editors wasn't happy with CRLF. I've reverted the line endings back to DOS

File: flaml/model.py (hunk 3 above)
Comment (markharley) [reply flaml/model.py]: Hey. Thanks for the review. Yeah, looks like one of our editors wasn't happy with CRLF. I've reverted the line endings back to DOS

File: website/docs/Use-Cases/Task-Oriented-AutoML.md (hunk 4 above)
Comment (markharley) [reply website/docs/Use-Cases/Task-Oriented-AutoML.md]: Hey! Thanks for the review.  Yeah, this and a couple of other instances appear to be a mistaken replacement. Likely Pycharm refactor going over-the-top. I've fixed them up now

**Comment by markharley** [comment]
> Was looking at the failed build and noticed that there were some import issues.
> 
> * Python 3.6 doesn't include data classes, but there is a backport that could be added to the build script (and setup).
> * Also saw that importing holidays fails. That's not part of the [test] requirements in setup.py (see https://github.com/markharley/FLAML/blob/fafd6524c9466d8ab9c9049bb6102c34cd5819ec/setup.py#L91).
> * It sounds like new tests were added for time series tasks? Should the requirements in [ts_forecast] and [forecast] also be included in [test] in that case?

Hey! Thanks for the review. I've added dataclasses as a Python version dependent install requirement, thanks for spotting it! Holidays turned out not to be used outside of the tests at present so I've removed the unnecessary test in test_forecast

**Comment by markharley** [comment]
> > Was looking at the failed build and noticed that there were some import issues.
> > 
> > * Python 3.6 doesn't include data classes, but there is a backport that could be added to the build script (and setup).
> > * Also saw that importing holidays fails. That's not part of the [test] requirements in setup.py (see https://github.com/markharley/FLAML/blob/fafd6524c9466d8ab9c9049bb6102c34cd5819ec/setup.py#L91).
> > * It sounds like new tests were added for time series tasks? Should the requirements in [ts_forecast] and [forecast] also be included in [test] in that case?
> 
> The holidays are only used in time series forecasting which is optional. Make sure the import statement is invoked only under the time series environment. See https://github.com/microsoft/FLAML/blob/main/flaml/automl.py#L1096 as an example for nlp (another optional environment).

Hey! Thanks for the review @liususan091219. I've applied your relative import suggestions. On `holidays`, it turns out that it wasn't called outside of tests at present and so I've removed the offending test and redundant code paths

**Comment by markharley** [comment]
> Would it be more reasonable to put the "nlp" and "time_series" folders into the "automl" folder?

Hey! I think we ended up with this layout to avoid some circular imports, but I'll have another try at refactoring these into the automl subpackage :smile:

**Comment by sonichi** [comment]
Some checks failed. I wonder if it will be easier to break the PR down to smaller PRs. For example, the first PR to make is to just create the `automl` subpackage and move files to corresponding locations. Otherwise this PR would take a long time to merge and you will get conflicts often.

**Comment by qingyun-wu** [comment]
> 


Thank you @markharley! We indeed need to re-organize the whole structure of the **flaml** folder considering the need of adding an automl folder. I am attaching a proposal for the new structure (considering all content in flaml, not just the changes involved in this PR). 

In this PR, perhaps you can just make .py files about automl and the time_series folder in the right place. We can come up with a plan with the other changes (and perhaps also discuss this proposed structure plan in the maintainer meeting on 10/10). 

- **automl**
     - **nlp**
     - **time_series** 
      automl.py
      data.py
      ml.py
      model.py
      train_log.py
      config.py
      [and other files need to be added in the automl folder]
- **tune**
    - **searcher**
    - **scheduler**
    [and existing files in the tune foler]
- **onlineml**
- **default**

File: flaml/automl/task.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+class Task(ABC):
+    estimators = {}
+
+    def __init__(
+        self,
+        task_name: str,
+        X_train: Union[np.ndarray, pd.DataFrame],
+        y_train: Union[np.ndarray, pd.DataFrame, pd.Series],
+    ):
```
Comment (sonichi) [inline flaml/automl/task.py:66]: The constructor here indicates that a Task object needs to know about X_train and y_train when it's constructed. 
The implementation indicates that no reference to the dataset is stored inside the Task object.
Why? What's the relation between a Task object and a dataset exactly?

Comment (EgorKraevTransferwise) [reply flaml/automl/task.py:66]: for example, at this stage we could infer it it's a binary or multi-category classification, and whether it's a univariate or panel regression, so the user wouldn't have the hassle of specifying that

File: flaml/automl/task.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        objective_name = "multiclass"
+    return objective_name
+
+
+class Task(ABC):
+    estimators = {}
+
+    def __init__(
+        self,
+        task_name: str,
```
Comment (qingyun-wu) [inline flaml/automl/task.py:63]: Needs documentation on the allowed task_name name?

File: setup.py [hunk 14]
```diff
@@ -20,6 +20,7 @@
     "scipy>=1.4.1",
     "pandas>=1.1.4",
     "scikit-learn>=0.24",
+    "dataclasses>=0.8 ; python_version=='3.6'",
```
Comment (qingyun-wu) [inline setup.py:23]:
```suggestion
    "dataclasses>=0.8 ; python_version>='3.6'",
```
